src/data/real/dedekind_cut.lean

```diff
@@ -0,0 +1,403 @@
+/-
+Copyright (c) 2022 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import algebra.order.archimedean
+import data.set.pointwise
+import order.upper_lower
+
+/-!
+# Dedeking cuts
+
+A basic theory of Dedekind cuts, used in a construction of the reals.
+
+## Important definitions
+
+- `dedekind_cut Q` is an `add_comm_group` when `Q` is `linear_ordered_field` and `archimedean`
+- TODO: finish constructing the full field
+
+## Tags
+
+dedekind, real, cut
+
+## References
+
+[Rudin, *Principles of Mathematical Analysis* (Chapter 1, Appendix)][rudin1976]
+
+-/
+
+section
+
+variables (Q : Type*) [has_lt Q]
+
+/-- *Dedekind cut*, defined for an arbitrary type with a less-than relation -/
+@[ext] structure dedekind_cut :=
+(carrier : set Q)
+(nonempty' : carrier.nonempty) -- Step 1, (I), first half
+(ne_top' : carrier ≠ ⊤) -- Step 1, (I), second half
+(is_lt_lower' : ∀ ⦃p q : Q⦄, q < p → p ∈ carrier → q ∈ carrier) -- Step 1, (II)
+(exists_gt' : ∀ ⦃p : Q⦄, p ∈ carrier → ∃ (r : Q) (hr : r ∈ carrier), p < r) -- Step 1, (III)
+
+variables {Q}
+
+namespace dedekind_cut
+
+instance : set_like (dedekind_cut Q) Q :=
+⟨dedekind_cut.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+variables {α β γ : dedekind_cut Q} {p q r : Q}
+
+@[simp]
+lemma mem_carrier : p ∈ α.carrier ↔ p ∈ α := iff.rfl
+
+@[simp]
+lemma mem_coe : p ∈ (α : set Q) ↔ p ∈ α := iff.rfl
+
+@[simp]
+lemma mem_mk {s : set Q} (hn hne hl hg) : p ∈ mk s hn hne hl hg ↔ p ∈ s := iff.rfl
+
+variables (α)
+
+lemma nonempty : (α : set Q).nonempty := α.nonempty'
+lemma lt_top : (α : set Q) < ⊤ :=
+begin
+  split,
+  { simp },
+  have := α.ne_top',
+  contrapose! this,
+  ext,
+  simp only [mem_carrier, set.top_eq_univ, set.mem_univ, iff_true],
+  exact this (by simp)
+end
+lemma is_lt_lower ⦃p q : Q⦄ (h : q < p) (hp : p ∈ α) : q ∈ α := α.is_lt_lower' h hp
```

Comment:
Not a fan of having all the lemmas smushed together, please separate them unless they're very short and very related.

Reply:
They are very related -- they are restatements of the axioms of a cut, just with different binders.

Reply:
But they're not very short. Unless dealing with one-liner lemmas, I find that doing this greatly hampers readability -- it's hard to tell by glance where one lemma ends and another beings.

Reply:
Instead of just adding spacing, I've added an actual comment explaining what the lemmas are for. In fact, these are not lemmas that should be "read", and I am using the spacing (lack thereof) to indicate that. Hopefully, the comment will help.